src/quota_aggregator_impl.cc

```diff
@@ -0,0 +1,255 @@
+  AllocateQuotaRequest request;
+
+  request.set_service_name(service_name);
+  request.set_service_config_id(service_config_id);
+
+  if (operation_aggregator_ != NULL) {
+    *(request.mutable_allocate_operation()) =
+        operation_aggregator_->ToOperationProto();
+    operation_aggregator_ = NULL;
+  }
+
+  return request;
+}
+
+QuotaAggregatorImpl::QuotaAggregatorImpl(const std::string& service_name,
+                                         const std::string& service_config_id,
+                                         const QuotaAggregationOptions& options)
+    : service_name_(service_name),
+      service_config_id_(service_config_id),
+      options_(options) {
+  if (options.num_entries > 0) {
+    cache_.reset(new QuotaCache(
+        options.num_entries, std::bind(&QuotaAggregatorImpl::OnCacheEntryDelete,
+                                       this, std::placeholders::_1)));
+    cache_->SetAgeBasedEviction(options.refresh_interval_ms / 1000.0);
+  }
+}
+
+QuotaAggregatorImpl::~QuotaAggregatorImpl() {
+  SetFlushCallback(NULL);
+  FlushAll();
+}
+
+// Sets the flush callback function.
+// The callback function must be light and fast.  If it needs to make
+// a remote call, it must be non-blocking call.
+// It should NOT call into this object again from this callback.
+// It will cause dead-lock.
+void QuotaAggregatorImpl::SetFlushCallback(FlushCallback callback) {
+  InternalSetFlushCallback(callback);
+}
+
+// If the quota could not be handled by the cache, returns NOT_FOUND,
+// caller has to send the request to service control.
+// Otherwise, returns OK and cached response.
+::google::protobuf::util::Status QuotaAggregatorImpl::Quota(
+    const ::google::api::servicecontrol::v1::AllocateQuotaRequest& request,
+    ::google::api::servicecontrol::v1::AllocateQuotaResponse* response) {
+  if (request.service_name() != service_name_) {
+    return Status(Code::INVALID_ARGUMENT,
+                  (string("Invalid service name: ") + request.service_name() +
+                   string(" Expecting: ") + service_name_));
+  }
+
+  if (!request.has_allocate_operation()) {
+    return Status(Code::INVALID_ARGUMENT,
+                  "allocate operation field is required.");
+  }
+
+  if (!cache_) {
+    // By returning NO_FOUND, caller will send request to server.
+    return Status(Code::NOT_FOUND, "");
+  }
+
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer stack_buffer(this);
+  MutexLock lock(cache_mutex_);
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer::Swapper swapper(
+      this, &stack_buffer);
+
+  string request_signature = GenerateAllocateQuotaRequestSignature(request);
+  QuotaCache::ScopedLookup lookup(cache_.get(), request_signature);
+
+  if (!lookup.Found()) {
+    // Create a temporary element which will be remained in the cache until
+    // actual response arrives
+    ::google::api::servicecontrol::v1::AllocateQuotaResponse temp_response;
+    InternalCacheResponse(request, temp_response, true);
```

Comment:
for this call, it is different than other case, 
== you have signature, 
== you know it is NOT in the cache. 
so don't re-use the function.